### PR TITLE
HYDROPHONES: Processing Re-Write

### PR DESCRIPTION
**Pull request by MarshallRawson**
See the docs for how to use

The [Triggering Time Results](https://docs.google.com/spreadsheets/d/16HoXAi3OgLhK5WiuQbdKxn1dbtSwVf53f8DfPNMiTVQ/edit?usp=sharing)

NOTE: There is one Ping my code triggers on that Forrest's code doesnt, but when stepping throught and looking at the waveform, I believe that it is an actual ping that must have been to faint for the Sylphase code

NOTE: In my code I walk out 0.0003 both in front and behind the time of triggering, and all the errors between my code and Forrest's code seems to be within that margin.

### CODE REVIEW

File: mil_common/drivers/mil_passive_sonar/src/sylphase_ros_bridge.cpp [hunk 1]
```diff
@@ -109,7 +106,7 @@ void SylphaseSonarToRosNode::read_messages(boost::asio::ip::tcp::socket& socket)
       // Received packets are Big-Endian, convert to system
       for (size_t i = 0; i < msg.data.size(); ++i)
       {
-        msg.data[i] = ntohs(msg.data[i]) + SIGNED_TO_UNSIGNED;
+        msg.data[i] = ntohs(msg.data[i]);
```
Comment (kev-the-dev) [inline mil_common/drivers/mil_passive_sonar/src/sylphase_ros_bridge.cpp]: The `Ping.msg` is UNSIGNED whereas the data from sylphase is signed.... so I think this will cause problems as random signed intergers will be interpreted as unsigned on the receiving end.

So either:
* Revert this change
* Change the data type in `Ping.msg` to int16

File: mil_common/utils/mil_msgs/srv/Disable.srv [hunk 2]
```diff
@@ -0,0 +1,2 @@
+---
```
[the remark this answers is not in the document]
Comment (MarshallRawson) [reply mil_common/utils/mil_msgs/srv/Disable.srv]: Good Point, up to date as of 
`TOOLS: Made mil ros plotter std_srv enable, removed Enable Disable srv`

File: mil_common/drivers/mil_passive_sonar/scripts/triggering [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+from scipy.signal import remez, fftconvolve, correlate
+from scipy.ndimage.filters import maximum_filter1d
+
+
+class HydrophoneInterpreter:
+
+    def __init__(self):
+        #  Ros stuff
+        rospy.init_node('triggering', log_level=rospy.DEBUG)
+        self.sub = rospy.Subscriber('/samples', numpy_msg(Ping), self.hydrophones_cb)
```
[the remark this answers is not in the document]
Comment (MarshallRawson) [reply mil_common/drivers/mil_passive_sonar/scripts/triggering]: I have added a HydrophoneSamples message as of Added HydrophoneSamples msg

File: mil_common/drivers/mil_passive_sonar/src/sylphase_ros_bridge.cpp (hunk 1 above)
Comment (MarshallRawson) [reply mil_common/drivers/mil_passive_sonar/src/sylphase_ros_bridge.cpp]: Good catch, Turns out this was causing some issues, Has been resolved as of Added HydrophoneSamples msg.

File: mil_common/utils/mil_tools/mil_ros_tools/mil_ros_plotter.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+from mil_msgs.srv import Enable, EnableResponse
+from mil_msgs.srv import Disable, DisableResponse
+import cv2
+from cv_bridge import CvBridge
+from matplotlib.backends.backend_agg import FigureCanvasAgg
+from matplotlib.figure import Figure
+import matplotlib
+import threading
+
+class MilRosPlotter:
```
[the remark this answers is not in the document]
Comment (MarshallRawson) [reply mil_common/utils/mil_tools/mil_ros_tools/mil_ros_plotter.py]: Added as of `TOOLS: Plotter docstring added`

File: mil_common/drivers/mil_passive_sonar/scripts/triggering [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        time = time[(window_size-1): data.shape[0]]
+        # take the gradient of the max convolution (we are looking for steep increase = big gradient)
+        gradient = np.gradient(max_convolve, axis=0)
+
+        triggered_at_idx = False
+        if np.max(gradient >= self.threshold):
+            print 'triggered at ', time[np.argmax(gradient)]
+            triggered_at_idx = np.argmax(gradient)
+
+        if self.trigger_debug.enabled:
```
[the remark this answers is not in the document]
Comment (MarshallRawson) [reply mil_common/drivers/mil_passive_sonar/scripts/triggering]: Good idea, implimented as of  `TOOLS: Plotter: docstring added, lazy publisher`

File: mil_common/drivers/mil_passive_sonar/scripts/triggering [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+from mil_passive_sonar.msg import Ping
+from mil_ros_tools.mil_ros_plotter import MilRosPlotter
+
+
+from scipy.fftpack import rfft
+from scipy.signal import remez, fftconvolve, correlate
+from scipy.ndimage.filters import maximum_filter1d
+
+
+class HydrophoneInterpreter:
```
[the remark this answers is not in the document]
Comment (MarshallRawson) [reply mil_common/drivers/mil_passive_sonar/scripts/triggering]: Fixed as of `HYDROPHONES: Triggering: added target frequency`

File: mil_common/utils/mil_tools/mil_ros_tools/plotter.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+               self.pub.get_num_connections() and\
+               (self.thread is None or not self.thread.is_alive())
+
+
+    def publish_plots(self, plots, titles=[]):
+        if self.pub.get_num_connections() == 0:
+            return
+        if not self.enabled:
+            return
+        if (self.thread is not None) and self.thread.is_alive():
```
Comment (kev-the-dev) [inline mil_common/utils/mil_tools/mil_ros_tools/plotter.py]: All of these lines could be replaced with
```
if not self.is_go():
  return
```

File: mil_common/drivers/mil_passive_sonar/scripts/triggering [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+
+    def enable(self, req):
+        self.enabled = req.data
+        res = SetBoolResponse()
+        res.success = True
+        return res
+
+
+    def set_target(self, req):
```
Comment (kev-the-dev) [inline mil_common/drivers/mil_passive_sonar/scripts/triggering]: So setting the target frequency doesn't affect the band pass? Wasn't the whole point of the redesign to have the bandpass change based on the freqency we're listening to?

File: mil_common/drivers/mil_passive_sonar/src/sylphase_ros_bridge.cpp [hunk 9]
```diff
@@ -136,7 +133,7 @@ void SylphaseSonarToRosNode::read_messages(boost::asio::ip::tcp::socket& socket)
 
 int main(int argc, char** argv)
 {
-  ros::init(argc, argv, "my_node_name");
+  ros::init(argc, argv, "sylphase_ros_bridge");
```
Comment (kev-the-dev) [inline mil_common/drivers/mil_passive_sonar/src/sylphase_ros_bridge.cpp]: my oof

File: mil_common/drivers/mil_passive_sonar/scripts/triggering [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+        self.enabled = True
+
+        # Attriburtes about our general frequency range (all pinger live here)
+        #  Frequency range garunteed to be relatively quiet except for the pingers (in Hz)
+        self.general_lower = 15000 # lowest frequency pinger - 10 kHz
+        self.general_upper = 50000 # highest frequency pinger + 10 kHz
+
+        # Attributes about our target frequency range
+        #  target Frquency in Hz
+        self.target = rospy.get_param('~target_frequency')
```
Comment (kev-the-dev) [inline mil_common/drivers/mil_passive_sonar/scripts/triggering]: So we dropped this being configurable via a service? We will want missions to be able to on the fly change this without restarting this node. I could imagine it being an alias like `set_pinger_freq 25000`

File: SubjuGator/command/sub8_launch/config/passive_sonar.yaml [hunk 11]
```diff
@@ -1,3 +1,18 @@
...
-sub8_hydrophones/v_sound            :  1482
+# distance from any hydrophone to hydrophone 0
+dist_h: 2.286e-02
+# speed of sound in water
+v_sound: 1482
+#  target Frquency in Hz
+triggering/target_frequency: 25000
+#  tolerance around target frequerncy in Hz
+triggering/frequency_tolerance: 100
+# how many terms the filter has, dont touch
```
[the remark this answers is not in the document]
Comment (MarshallRawson) [reply SubjuGator/command/sub8_launch/config/passive_sonar.yaml]: fixed (and added some other important ones)

File: docker/dev/user_install [hunk 12]
```diff
@@ -27,7 +27,19 @@ mil_user_install_dependencies()
...
-    nmap
+    nmap \
+    xvfb \
+    xdg-utils \
+    libappindicator3-1 \
+    libsecret-1-0
+
+  # install drawio
+  sudo pip3 install \
+   sphinxcontrib-drawio
```
[the remark this answers is not in the document]
Comment (MarshallRawson) [reply docker/dev/user_install]: These have been moved to system install

File: mil_common/drivers/mil_passive_sonar/msg/Ping2.msg [hunk 13]
```diff
@@ -0,0 +1,9 @@
+Header header
+int32 channels # Number of channels
+int32 samples # Number of samples on each channel (length of data should be sampples * channels)
+int32 sample_rate # Number of samples / second recorded by a channel 
+#   all channels must use the same sample rate
+int16[] data # data relevent to the ping that was triggered
```
[the remark this answers is not in the document]
Comment (MarshallRawson) [reply mil_common/drivers/mil_passive_sonar/msg/Ping2.msg]: what do you think about just adding ping time to hydrophone samples message and only using / populating it when it makes sense?

File: mil_common/drivers/mil_passive_sonar/scripts/triggering (hunk 10 above)
Comment (MarshallRawson) [reply mil_common/drivers/mil_passive_sonar/scripts/triggering]: run time configuration now supported, see docs

File: mil_common/drivers/mil_passive_sonar/src/sylphase_ros_bridge.cpp (hunk 9 above)
Comment (MarshallRawson) [reply mil_common/drivers/mil_passive_sonar/src/sylphase_ros_bridge.cpp]: np

File: mil_common/drivers/mil_passive_sonar/scripts/ping_locator [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        self.debug_samples.publish_plots(plots, titles, vlines)
+
+        try:
+            vec = util.calculate_dir_pinger(deltas, self.dist_h, self.v_sound)
+        except Exception as e:
+            rospy.logwarn("/hydrophones/ping_locator could not calculate pinger direction %s"%e)
+            return
+
+        v = Vector3Stamped()
+        v.header = Header()
```
[the remark this answers is not in the document]
Comment (MarshallRawson) [reply mil_common/drivers/mil_passive_sonar/scripts/ping_locator]: good catch, has be resolved

**Comment by MarshallRawson** [comment]
> Excellent documentation and a clear design improvement from our existing system. Noticed a couple things in my initial glance at it.
> 
> Also curious to see the validation using the bags collected with labeled ground truth angles

Have tested with the bags from most recent test day, seems to work great at 10ft in all directions. However, at 3ft, alot fewer pings get all the way through to the direction marker (although the ones that do are almost all correct). I am not sure, but I think it has to do with more echos happening the closer you get to the pinger, especially off the floor. Also I noticed a weird thing happening to the data where the samples appear to have a discontinuity, I added a way to filter them out (they were producing bad results), but I would like to talk to Forrest about it.

File: mil_common/drivers/mil_passive_sonar/scripts/triggering [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+        #  max convolution window, the amount of time that we assume should be quiet before a ping in sec
+        self.window_time = 0.1
+        #  prev data buffer
+        self.prev_data = None
+
+        # Ros stuff
+        rospy.init_node('hydrophone_trigger')
+        # if running with old bags use:
+        self.sub = rospy.Subscriber('samples', numpy_msg(Ping), self.hydrophones_cb)
+        self.sub = rospy.Subscriber('samples', numpy_msg(HydrophoneSamples), self.hydrophones_cb)
```
Comment (kev-the-dev) [inline mil_common/drivers/mil_passive_sonar/scripts/triggering]: ?

File: mil_common/drivers/mil_passive_sonar/msg/Ping2.msg (hunk 13 above)
[the remark this answers is not in the document]
Comment (kev-the-dev) [reply mil_common/drivers/mil_passive_sonar/msg/Ping2.msg]: It seems that these are different purposes and should have different messages. Nesting them is the most appropriate

**Comment by MarshallRawson** [comment]
> p good but still has the confusing message names

Take a look at most recent push, I have restructured the passive sonar messages to be cleaner

File: mil_common/drivers/mil_passive_sonar/scripts/triggering (hunk 15 above)
Comment (MarshallRawson) [reply mil_common/drivers/mil_passive_sonar/scripts/triggering]: See most recent commit, I have changed how the support for Ping message type works

**Comment by MarshallRawson** [comment]
> Didn't totally re-review, but looks like the message names are now reasonable. I'd add an issue about deprecating the old stuff, to do later



@kev-the-dev  what do you think about holding off completely removing the old stuff until we get in the water again?

Also, the force push was just me rebasing on master

**Comment by MarshallRawson** [comment]
> > Didn't totally re-review, but looks like the message names are now reasonable. I'd add an issue about deprecating the old stuff, to do later
> 
> @kev-the-dev what do you think about holding off completely removing the old stuff until we get in the water again?
> 
> Also, the force push was just me rebasing on master

Also, I have verified that it functions after the rebase.
